src/IR/Constraint/Constrain/Expression.hs

```diff
@@ -1,50 +1,54 @@
-import           Control.Monad                  ( zipWithM )
-import           Data.Foldable                  ( foldrM )
-import qualified Data.Map.Strict               as Map
-import qualified IR.Constraint.Canonical       as Can
-import           IR.Constraint.Canonical        ( (-->) )
-import qualified IR.Constraint.Constrain.Annotation
-                                               as Ann
-import qualified IR.Constraint.Constrain.Pattern
-                                               as Pattern
-import           IR.Constraint.Monad            ( TC
-                                                , freshVar
-                                                , freshAnnVar
-                                                , getExtern
-                                                , throwError
-                                                )
-import           IR.Constraint.Type            as Type
-import qualified IR.IR                         as I
-import           IR.SegmentLets                 ( segmentDefs' )
+import qualified Common.Identifiers as Ident
+import Control.Monad (forM, zipWithM)
+import Data.Foldable (foldrM)
+import qualified Data.Map.Strict as Map
+import IR.Constraint.Canonical ((-->))
+import qualified IR.Constraint.Canonical as Can
+import qualified IR.Constraint.Constrain.Annotation as Ann
+import qualified IR.Constraint.Constrain.Pattern as Pattern
+import IR.Constraint.Monad (TC, freshAnnVar, getExtern, throwError)
+import IR.Constraint.Type as Type
+import qualified IR.IR as I
+import IR.SegmentLets (segmentDefs')
 
 
 -- | HELPER ALIASES
+type Def = (I.VarId, [I.Annotation], Variable, I.Expr Attachment)
 
 
-type Attachment = (I.Annotations, Variable)
-
-type BinderDef = (I.Binder, I.Expr Attachment)
-type Def = (I.VarId, I.Expr Attachment)
+type BinderDef = (I.Binder Attachment, I.Expr Attachment)
 
 
 -- | DEFS
-
-
 constrainBinderDefs :: [BinderDef] -> Constraint -> TC Constraint
 constrainBinderDefs binderDefs finalConstraint = do
-  defs <- mapM
-    (\(binder, expr) -> do
-      name <- binderToVarId binder
-      return (name, expr)
-    )
-    binderDefs
+  defs <- forM binderDefs \(binder, expr) -> do
+    -- TODO: these binders will also contain annotations; we need to do
+    -- something with them!
+
+    -- resolved I think?
+    name <- Type.binderToVarId binder
+    let (anns, uvar) = uncarryAttachment binder
+    return (name, anns, uvar, expr)
   constrainDefs defs finalConstraint
 
 
+-- consDefs <- constrainDefs defs finalConstraint
+-- let annsVarPairs = map (uncarryAttachment . fst) binderDefs
+--     -- vars = map snd annsVarPairs
+--     exprTyps = map (TVarN . snd . uncarryAttachment . snd) binderDefs
+--     forBinders ((ann, var), exprTyp) =
+--       exists [var] <$> do
+--         Ann.withAnnotations ann (TVarN var) \varExp ->
+--           return $ CEqual varExp exprTyp
+-- constraints <- mapM forBinders (zip annsVarPairs exprTyps)
+-- return $ CAnd (consDefs : constraints)
+
```

Comment:
I forgot to remove this commented-out block of code when I committed my fix. Could you help me clean this up @yiming-fang

Reply:
Other than this, I have no other suggestions. This PR is looking good!

Reply:
> I forgot to remove this commented-out block of code when I committed my fix. Could you help me clean this up @yiming-fang

Done!